File: 2023/python/day01b/day01b.py
#!/usr/bin/env python3

import sys

def usage():
    print(f'usage: {sys.argv[0]} <file>')
    sys.exit(1)

def process(contents):
    lines = contents.split()
    nums = []
    digits_map = { '0': '0', '1': '1', '2': '2', '3': '3', '4': '4', '5': '5',
                   '6': '6', '7': '7', '8': '8', '9': '9', 'zero': '0',
                   'one': '1', 'two': '2', 'three': '3', 'four': '4',
                   'five': '5', 'six': '6', 'seven': '7', 'eight': '8',
                   'nine': '9' }
    for line in lines:
        left_indices = {}
        right_indices = {}
        for digit in digits_map.keys():
            if digit in line:
                left_indices[digit] = line.find(digit)
                right_indices[digit] = line.rfind(digit)
        left_digit = min(left_indices, key=left_indices.get)
        right_digit = max(right_indices, key=right_indices.get)
        nums.append(int(digits_map[left_digit] + digits_map[right_digit]))
    return sum(nums)

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    try:
        infile = open(filename)
        contents = infile.read()
        result = process(contents)
        print(f'result = {result}')
    except IOError:
        print(f'read of input file "{filename}" failed.')
        sys.exit(1)

if __name__ == '__main__':
    main()

File: 2023/cpp/day01b/day01b.cpp
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

void usage(char* progname)
{
    std::cout << progname << " <file>" << std::endl;
    std::exit(1);
}

int process(std::ifstream& infile)
{
    std::string line;
    std::unordered_map<std::string, std::string> digits = {
        {"0", "0"}, {"1", "1"}, {"2", "2"}, {"3", "3"}, {"4", "4"},
	{"5", "5"}, {"6", "6"}, {"7", "7"}, {"8", "8"}, {"9", "9"},
	{"zero", "0"}, {"one", "1"}, {"two", "2"}, {"three", "3"},
	{"four", "4"}, {"five", "5"}, {"six", "6"}, {"seven", "7"},
	{"eight", "8"}, {"nine", "9"},
    };
    std::vector<int> nums;
    while ( std::getline(infile, line) )
    {
        std::unordered_map<std::string, int> left_indices;
        std::unordered_map<std::string, int> right_indices;
        for ( auto digitp = digits.begin(); digitp != digits.end(); digitp++ )
        {
            if ( line.find(digitp->first) != std::string::npos )
            {
                left_indices.insert({digitp->first, (int)line.find(digitp->first)});
                right_indices.insert({digitp->first, (int)line.rfind(digitp->first)});
            }
        }
        int min_index = (int)line.size() + 1;
        std::string left_digit;
        for ( auto leftp = left_indices.begin(); leftp != left_indices.end(); leftp++ )
        {
            if ( leftp->second < min_index )
            {
                min_index = leftp->second;
                left_digit = digits[leftp->first];
            }
        }
        int max_index = -1;
        std::string right_digit;
        for ( auto rightp = right_indices.begin(); rightp != right_indices.end(); rightp++ )
        {
            if ( rightp->second > max_index )
            {
                max_index = rightp->second;
                right_digit = digits[rightp->first];
            }
        }
        std::string num_str = left_digit + right_digit;
        nums.push_back(std::atoi(num_str.c_str()));
    }
    int sum = 0;
    for ( auto nump = nums.begin(); nump != nums.end(); nump++ )
    {
        sum += *nump;
    }
    return sum;
}

int main(int argc, char* argv[])
{
    if ( argc < 2 ) {
        usage(argv[0]);
    }
    std::ifstream infile(argv[1]);
    int result = process(infile);
    std::cout << "result = " << result << std::endl;
    return 0;
}
